### ISSUE TITLE
JS module for generating SVG?

### ISSUE THREAD

**Issue by dbarvitsky** [umlet/umlet#394]
First off, I would like to thank you for all your work. IMHO this is one of the best UML authoring tools out there. I am an architect for fun and profit for the past 12 years; and decent UML tools are hard to find. I started with Visio with UML templates by www.softwarestencils.com (not bad, but difficult to collaborate because people don't like to pay for Visio), PUML (great concept, but no way to change layout easily, and dot/neato does not always generate appealing layouts), lucidchart (flashy, but takes a lot of clicks to get job done) and a bunch of others. I recently tried Umlet and I am a huge fan now - authoring is fast, the tool is free and diagrams look good. I have one request though.

Here is what I am trying to do for my folks.  We store requirements and designs in Git, together with source code, so that we can version it with source and keep it (as much as possible) up-to-date.  I created a simple NodeJS app that renders markdown and diagrams. This way we can edit the requirements, fire the build and my guys can see all architect's fantasies in their browser. We can do diffs / merges and all stuff. The app generates content on the fly, so I can edit things and see end-results in real time too. Pretty neat, huh?

Obviously diagrams are part of it. And here comes the problem. Currently we have to store SVGs in source control (yuck!) and remember to update them. Which rarely happens because people suck at remembering anything. 

TLDR: Hence, I would like to generate SVGs from UXFs on the fly in the app, preferably in JavaScript.

I looked at your source, looks like MainView#onExportMenuItemClick does what I need (DiagramXmlParser+CanvasUtils for SVG generation?). I never worked with GWT, but I believe there should be a way of generating JS-callable code with it. Wrapping this into a NodeJS module, I think, would solve my problem.

So, here is the proposal. I'd be happy to pay, say, $100 for this feature to support your project, if you are willing to take on this. Or I can try to do it myself and contribute whatever comes out of it, in which case any pointers would be much appreciated).

Thanks!

**Comment by afdia** [comment]
Hi
That's an interesting feature request, but I guess it is not possible by simply using GWT, because GWT simulates only a minor portion of the Java API (see http://www.gwtproject.org/doc/latest/RefJreEmulation.html) and the whole AWT and Swing package are not part of it, therefore the SVG export library will not work.

But as you may have seen there is already a GWT based version of UMLet (http://www.umletino.com/) which is compiled from the sources in the umlet-gwt module (umlet-elements contains the codebase which is used in all versions of UMLet and therefore limited to the API which is emulated by GWT)
To make the core application work with Swing and GWT, we have already created an abstraction layer (the DrawHandler.java) which can be implemented multiple times (currently there is the awt.Graphics2D implementation for UMLet and the HTML Canvas based implementation for Umletino)
Here is the GWT implementation: https://github.com/umlet/umlet/blob/master/umlet-gwt/src/main/java/com/baselet/gwt/client/element/DrawHandlerGwt.java
Unfortunately a canvas s bitmap based and therefore cannot be simply exported to SVG, but you may try to implement a separate DrawHandler to redirect the drawing calls to a SVG library.

I have already created issue #245 some time ago which is basically what you need (implementing an SVG export for Umletino). Perhaps you can try those libraries and see if they work.

I guess it shouldn't be that hard, because as mentioned before the abstraction layer (DrawHandler) already exists, the missing part is the SVG based implementation for GWT.
Unfortunately I do not have much time to work on Umlet at the moment, but if you want to give it and contribute the results we would be happy to include the svg implementation of the DrawHandler.

**Comment by dbarvitsky** [comment]
Hi. Thanks for the pointers! I looked at the DrawHandlerGwt.java, it looks pretty straightforward. Worst case I figured I can come up with a Java binary that does the conversion, but I'll look into other options. Thanks for suggestions!

**Comment by afdia** [comment]
if the GWT approach doesnt work you can also use the umlet.jar from the command line.

e.g. (I cannot test the call here, but it should be correct):
java -jar umlet.jar -action=convert -format=svg -filename=input.uxf -output=D:\output.svg
